### DISCUSSION TITLE
Custom CSS styling to Theia's top panel

### DISCUSSION THREAD

**Discussion by Abhay0294** [eclipse-theia/theia#11194] [Q&A] [answered]
I have implemented an additional top-panel above the default one in theia built tool, now I wanted to do some CSS styling to it like color change, height, width, etc. 
Also attached the snap of the implementation by pointing where the styling is needed
<img width="907" alt="Top-panle" src="https://user-images.githubusercontent.com/57217189/170206623-42ea0860-044c-4a92-800d-9ca2b17db4e9.PNG">
.

**Comment by msujew** [comment, marked as answer]
Hi @Abhay0294, is this a question?

Anyway, if you're trying to modify the CSS of the top panel, you can do so using the `theia-top-panel` id. Like this:

```css
#theia-top-panel {
  min-height: 100px;
  padding-top: 68px;
}
```

And then import your CSS in some part of your app, likely your frontend module:

```ts
import 'relative/path/to/file.css';
```

**Reply by Abhay0294** [reply]
Hi @msujew, sorry for that silly question.
The prblem which I am facing now is that since I have created the two top panels, both got the same id i.e. theia-top-panel so whenever I am applying any styling to the one the same is getting reflected to the second top-panel.

So, if possible can you suggest me the ways through which I can make them independent of each other.

Things which I have tried:
1. While overriding the application shell's createTopPanel function with the different id, then this id gets assigned to the original panel also. [The reason being the object reference concept.]

Below are the Shell's overriden codes:

1.createTopPanel function:
@postConstruct()
protected createTopPanel(): Panel {
const panel = super.createTopPanel();
const customPanel = panel;
customPanel.id = "new-theia-top-panel";
return customPanel;
}

2.createLayout function:

protected createLayout() {
const bottomSplitLayout = this.createSplitLayout(
[this.mainPanel, this.bottomPanel],
[1, 0],
{ orientation: 'vertical', spacing: 0 }
);
const panelForBottomArea = new SplitPanel({ layout: bottomSplitLayout });
panelForBottomArea.id = 'theia-bottom-split-panel';
const leftRightSplitLayout = this.createSplitLayout(
[this.leftPanelHandler.container, panelForBottomArea, this.rightPanelHandler.container],
[0, 1, 0],
{ orientation: 'horizontal', spacing: 0 }
);
const panelForSideAreas = new SplitPanel({ layout: leftRightSplitLayout });
panelForSideAreas.id = 'theia-left-right-split-panel';
return this.createBoxLayout(
[this.createTopPanel(), this.topPanel, panelForSideAreas, this.statusBar],
[0, 0, 1, 0],
{ direction: 'top-to-bottom', spacing: 0 }
);
}

**Reply by msujew** [reply]
Don't be sorry, I was joking, since there was no question mark, and it wasn't quite clear what you were exactly asking for help with :D

Anyway, there's an issue in your implementation. What happens is that your overriden `createTopPanel` method is called twice. The obvious one is your own `createLayout`. Theia will also call the method itself in its `init` routine and assign the result to `this.topPanel`:

https://github.com/eclipse-theia/theia/blob/0ec8075c1ef60cdbb0b1d4cdad40cbb44e21661b/packages/core/src/browser/shell/application-shell.ts#L269-L271

However, you want two separate `createTopPanel` methods, one which isn't overriden. Something like this:

```ts
export class CustomAppShell extends ApplicationShell {
    protected createLayout() {
        const bottomSplitLayout = this.createSplitLayout(
        [this.mainPanel, this.bottomPanel],
        [1, 0],
        { orientation: 'vertical', spacing: 0 }
        );
        const panelForBottomArea = new SplitPanel({ layout: bottomSplitLayout });
        panelForBottomArea.id = 'theia-bottom-split-panel';
        const leftRightSplitLayout = this.createSplitLayout(
        [this.leftPanelHandler.container, panelForBottomArea, this.rightPanelHandler.container],
        [0, 1, 0],
        { orientation: 'horizontal', spacing: 0 }
        );
        const panelForSideAreas = new SplitPanel({ layout: leftRightSplitLayout });
        panelForSideAreas.id = 'theia-left-right-split-panel';
        return this.createBoxLayout(
        [this.createSecondaryTopPanel(), this.topPanel, panelForSideAreas, this.statusBar],
        [0, 0, 1, 0],
        { direction: 'top-to-bottom', spacing: 0 }
        );
   }
   
   protected createSecondaryTopPanel() {
      const panel = super.createTopPanel();
      panel.id = "new-theia-top-panel";
      return panel ;
   }
}
```

**Comment by Abhay0294** [comment]
Thanks a lot @msujew for your help, I tried what you have suggested.

Outcomes after applying:
1. The secondaryTopPanel is still not visible.
2. After debugging got to know that top-panels with different ids we created as shown below from the inspect element snippet.
<img width="846" alt="Top-Panel-Issue" src="https://user-images.githubusercontent.com/57217189/170260475-4fcb15fe-8c2d-4518-a06f-5d49e650e4a6.PNG">

I think the class p-mod-hidden causing the issue. I also tried manipulating the css properties, due to which I am able to see its skeleton but still its not getting visible.

Please help me in this case if possible also suggest how this already initialsed classes(p-mod-hidden) can be removed

**Comment by Abhay0294** [comment]
**Update:** I am now able to get the two panels with two different ids. I just forgot to apply panel.show() after creating an object from super.createTopPanel(); , Since it is calling the parent's createTopPanel which returns the panel object with panel.hide().

Latest Overridable function:

protected createSecondaryTopPanel() {
const panel = super.createTopPanel();
console.log("Panel Object is:",panel)
panel.id = "new-theia-top-panel";
panel.show();
return panel;
}
